core/src/main/java/org/mule/runtime/core/MessagePropertiesContext.java

```diff
@@ -219,79 +247,62 @@ public String toString()
         return buf.toString();
     }
 
-    /**
-     * Check for properties that can't be serialized
-     */
-    private void writeObject(java.io.ObjectOutputStream out) throws IOException
+    private <T extends Serializable> T getValueOrDefault(T value, T defaultValue)
     {
-        for (Map.Entry<String, TypedValue> entry : inboundMap.entrySet())
+        //Note that we need to keep the (redundant) casts in here because the compiler compiler complains
+        //about primitive types being cast to a generic type
+        if (defaultValue == null)
         {
-            Object value = entry.getValue().getValue();
-            if (value != null && !(value instanceof Serializable))
-            {
-                String message = String.format(
-                    "Unable to serialize the %s message property %s, which is of type %s ",
-                    PropertyScope.INBOUND, entry.getKey(), value);
-                logger.error(message);
-                throw new IOException(message);
-            }
+            return value;
         }
-        for (Map.Entry<String, TypedValue> entry : outboundMap.entrySet())
+        else if (defaultValue instanceof Boolean)
         {
-            Object value = entry.getValue().getValue();
-            if (value != null && !(value instanceof Serializable))
-            {
-                String message = String.format(
-                    "Unable to serialize the %s message property %s, which is of type %s ",
-                    PropertyScope.OUTBOUND, entry.getKey(), value);
-                logger.error(message);
-                throw new IOException(message);
-            }
+            return  (T) (Boolean) ObjectUtils.getBoolean(value, (Boolean) defaultValue);
         }
-        out.defaultWriteObject();
-    }
-
-    private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException
-    {
-        in.defaultReadObject();
-    }
-
-    private static class UndefinedSessionPropertiesMap extends AbstractMap<String, TypedValue>
-        implements Serializable
-    {
-
-        private static final long serialVersionUID = -7982608304570908737L;
-
-        @Override
-        public Set<java.util.Map.Entry<String, TypedValue>> entrySet()
+        else if (defaultValue instanceof Byte)
         {
-            return Collections.emptySet();
+            return (T) (Byte) ObjectUtils.getByte(value, (Byte) defaultValue);
```

Comment:
If the above comment is done, can these type check be based on the TypedValue#dataType ?